libs/core/src/ports/port.ts

```diff
@@ -25,20 +25,106 @@ const log = logger(import.meta);
  */
 const adapters = new Map<string, Disposable>();
 
+/**
+ * Maps ports to default adapters aka factory names to adapter keys or names
+ */
+const defaultAdapters = new Map<string, string>();
+
+const InvalidKey = (key: string) => new Error(`Invalid adapter key: ${key}`);
+const ExistingDefaultAdapter = (key: string) =>
+  new Error(`Default adapter for ${key} port already exists`);
+
 /**
  * Wraps creation of adapters around factory functions
  * @param factory adapter of T factory function
  * @returns adapter instance
  */
 export function port<T extends Disposable>(factory: AdapterFactory<T>) {
-  return function (adapter?: T) {
-    if (!adapters.has(factory.name)) {
-      const instance = factory(adapter);
-      adapters.set(factory.name, instance);
-      log.info(`[binding adapter] ${instance.name || factory.name}`);
-      return instance;
+  return function (
+    options?:
+      | {
+          key: `${string}.${string}.${string}`;
+          adapter?: never;
+          isDefault?: never;
+        }
+      | {
+          key?: never;
+          adapter: T;
+          isDefault?: never;
+        }
+      | {
+          key: `${string}.${string}.${string}`;
+          adapter: T;
+          isDefault: boolean;
+        },
+  ) {
+    // If no option use default adapter or set in-memory adapter as default
+    if (!options) {
+      const defaultAdapterKey = defaultAdapters.get(factory.name);
+      if (defaultAdapterKey) return adapters.get(defaultAdapterKey) as T;
+
+      const adapterInstance = factory();
+      log.info(`[binding default adapter] ${adapterInstance.name}`);
+      defaultAdapters.set(factory.name, adapterInstance.name);
```

Comment:
I don't understand what is going on here. Assume you first pass in a default adapter without the key param. How does it generate the key name? If it is just factory.name, doesn't that not follow the `${string}.${string}.${string} pattern, so won't it fail on a subsequent key lookup?

Reply:
Only explicitly provided keys need to match the `string.string.string` pattern which is really only there for clarity. If you don't provide a key then that adapter cannot be fetched via a key thereafter. It is done this way to maintain the existing behaviour where the majority of port function invocation happens without any arguments.

So for example I can do
```
blobStorage({ adapter: S3BlobStorage() })

// storage = S3BlobStorage adapter
const storage = blobStorager();
```
Using a key for an adapter that is initialized without one is redundant because it requires passing around the adapter instance hence the `string.string.string` type which discourages this behaviour.
```
const s3Adapter = S3BlobStorage();
blobStorage({ adapter: s3Adapter })

// storage = S3BlobStorage adapter
const storage = blobStorager({ key: s3Adapter.name as any });
```